sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+    "substatus":"initialAssignment",
+    "lastUpdatedDateTimeUtc":"2020-04-10T03:11:13.2096201Z",
+    "etag":"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI="}}
+*/
+AZ_INLINE az_result az_iot_provisioning_client_payload_error_code_parse(
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorCode"), tm->name))
+  {
+    double value;
+    AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm->token, &value));
+    out_state->extended_error_code = (uint32_t)value;
+    out_state->status = (uint32_t)value / 1000;
+
+    return AZ_OK;
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+AZ_INLINE az_result az_iot_provisioning_client_payload_registration_state_parse(
+    az_json_parser* jp,
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (tm->token.kind != AZ_JSON_TOKEN_OBJECT_START)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  bool found_assigned_hub = false;
+  bool found_device_id = false;
+
+  while ((!(found_device_id && found_assigned_hub))
+         && az_succeeded(az_json_parser_parse_token_member(jp, tm)))
+  {
+    if (az_span_is_content_equal(AZ_SPAN_FROM_STR("assignedHub"), tm->name))
+    {
+      found_assigned_hub = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->assigned_hub_hostname));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("deviceId"), tm->name))
+    {
+      found_device_id = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->device_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorMessage"), tm->name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->error_message));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("lastUpdatedDateTimeUtc"), tm->name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->error_timestamp));
+    }
+    else if (tm->token.kind == AZ_JSON_TOKEN_OBJECT_START)
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(jp, tm->token));
+    }
+    else
+    {
+      (void)az_iot_provisioning_client_payload_error_code_parse(tm, out_state);
+    }
+  }
+
+  return AZ_OK;
+}
+
+AZ_INLINE az_result az_iot_provisioning_client_parse_payload(
+    az_span received_payload,
+    az_iot_provisioning_client_register_response* out_response)
+{
+  // Parse the payload:
+  az_json_parser jp;
+  az_json_token_member tm;
+  bool found_operation_id = false;
+  bool found_registration_state = false;
+  bool found_error = false;
+
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, received_payload));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_OBJECT_START)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  out_response->registration_information = registration_state_default();
+
```

Comment:
how about memset?

Reply:
the convention for things like this would be wise to use the default instead of memset. That way any changes to the default behavior gets updated here. Things aren't always guaranteed to be `0`.

Reply:
Well, agree, but we should design things that work in alignment with performance, right?

---
In reply to: [411636942](https://github.com/Azure/azure-sdk-for-c/pull/590#discussion_r411636942) [](ancestors = 411636942)

Reply:
Yea I don't necessarily disagree. If we are good with that tradeoff we can go for memset here.

Reply:
I don't think memset is appropriate here. The compiler can generate optimized code depending on what the dev selects (size, speed, etc).